codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -1176,18 +1195,35 @@ private void writePrefixHeaderDeserializerFunction(
             HttpBinding binding
     ) {
         String prefix = binding.getLocationName();
-        Shape targetValueShape = model.expectShape(targetShape.asMapShape().get().getValue().getTarget());
-        for (Shape shape : targetShape.asMapShape().get().members()) {
-            String name = shape.getId().getName();
-            String locationName = prefix + name;
-            writer.openBlock("if val := response.Header.Get($S); val != $S {",
-                    "}", locationName, "", () -> {
-                        String value = generateHttpBindingsValue(writer, model, symbolProvider,
-                                targetValueShape, binding, "val");
-                        writer.write("v.$L[$S] = $L", memberName, name,
-                                CodegenUtils.generatePointerValueIfPointable(writer, targetValueShape, value));
+
+        MemberShape valueMemberShape = targetShape.asMapShape()
+                .orElseThrow(() -> new CodegenException("prefix headers must target map shape"))
+                .getValue();
```

Comment:
We can also use         

`MapShape mapshape = model.expectShape(targetShape.getId(), MapShape.class);`

then you won't have to explicitly throw an exception. 

 I think even `targetShape.asMapShape().get()` would throw an exception directly.

Reply:
cool will switch to the expectShape call